### DISCUSSION TITLE
Custom Callbacks

### DISCUSSION THREAD

**Discussion by chirag-bajaj** [explosion/spaCy#11316] [Help: Coding & Implementations] [answered]
I want to add callbacks at the end of training to upload the best model to s3. I also want to update my model registry to update the training progress. 

Basically, I need a way to define custom callbacks at various stages in the pipeline and training. These callbacks would need access to loss, score, epochs, model saving path etc. 

What would be the best way to implement such a custom callback which has access to all these above-mentioned objects and more (in case the scope expands in the future)?

**Comment by thomashacker** [comment, marked as answer]
Hello,
You could create a custom logger which implements the callbacks. You can read more about how spaCy's loggers work [in the docs](https://spacy.io/api/top-level#loggers) and you can use the existing [ConsoleLogger.v1 ](https://github.com/explosion/spaCy/blob/master/spacy/training/loggers.py) as starting point.
